beacon/light/request/server.go

```diff
@@ -0,0 +1,215 @@
+}
+
+// LatestHead returns the server's latest reported head (slot and block root).
+// Note: the reason we can't return the full header here is that the standard
+// beacon API head event only contains the slot and block root.
+func (s *Server) LatestHead() (uint64, common.Hash) {
+	s.headLock.RLock()
+	defer s.headLock.RUnlock()
+
+	return s.latestHeadSlot, s.latestHeadHash
+}
+
+// canRequestNow returns true if a request can be sent to the server immediately
+// (has no timed out requests and underlying RequestServer does not require new
+// requests to be delayed). It also returns a priority value that is taken into
+// account when otherwise equally good servers are available.
+// Note: if canRequestNow ever returns false then it is guaranteed that a server
+// trigger will be emitted as soon as it becomes true again.
+func (s *Server) canRequestNow() (bool, uint64) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.isDelayed() || s.timeoutCount != 0 {
+		s.needTrigger = true
+		return false, 0
+	}
+	//TODO use priority based on in-flight requests (less is better)
+	return true, uint64(rand.Uint32() + 1)
+}
+
+// isDelayed returns true if the underlying RequestServer requires requests to be
+// delayed. In this case it also starts a timer to ensure that a server trigger
+// can be emitted when the server becomes available again.
+func (s *Server) isDelayed() bool {
+	delayUntil := s.RequestServer.DelayUntil()
+	if delayUntil == s.delayUntil {
+		return s.delayTimer != nil
+	}
+	if s.delayTimer != nil {
+		// Note: is stopping the timer is unsuccessful then the resulting AfterFunc
+		// call will just do nothing
+		s.stopTimer(s.delayTimer)
+		s.delayTimer = nil
+	}
+	s.delayUntil = delayUntil
+	delay := time.Duration(delayUntil - s.scheduler.clock.Now())
+	if delay <= 0 {
+		return false
+	}
+	s.delayTimer = s.scheduler.clock.AfterFunc(delay, func() {
+		if s.scheduler.testTimerResults != nil {
+			s.scheduler.testTimerResults = append(s.scheduler.testTimerResults, true) // simulated timer finished
+		}
+		s.lock.Lock()
+		if s.delayTimer != nil && s.delayUntil == delayUntil { // do nothing if there is a new timer now
+			s.delayTimer = nil
+			if s.needTrigger && s.timeoutCount == 0 {
+				s.needTrigger = false
+				s.scheduler.triggerServer(s)
+			}
+		}
+		s.lock.Unlock()
+	})
+	return true
+}
+
+// sendRequest generates a request ID and starts a timeout timer. If the timeout
+// is reached then the trigger is triggered (if specified).
+func (s *Server) sendRequest(timeoutTrigger *ModuleTrigger) uint64 {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	s.lastReqId++
+	reqId := s.lastReqId
+	s.timeouts[reqId] = s.scheduler.clock.AfterFunc(softRequestTimeout, func() {
+		if s.scheduler.testTimerResults != nil {
+			s.scheduler.testTimerResults = append(s.scheduler.testTimerResults, true) // simulated timer finished
+		}
+		s.lock.Lock()
+		if s.timeouts[reqId] != nil {
+			s.timeouts[reqId] = nil
```

Comment:
Why aren't you using `delete(s.timeouts, reqId)`?

Reply:
Added a comment to explain why not (nil means timed out, it is different from non existent entry).